Add auto calibrate (#19)
amperka/Octoliner: API.md

If the current sensor reading does not allow understanding of the line position (some trash under the sensor or gone out of track), the previous successful estimation (which is cached) is returned. The starting value is `0.0`.

### `uint8_t autoCalibrate(void)`
### `uint8_t autoCalibrate(enum CALIBRATE_MODE mode)`

Returns optimal sensitivity value for current octoliner/track pair. Mode can be: `CALIBRATE_MODE::DRAFT` (quick but can be not accuracy: about 2 sec), `CALIBRATE_MODE::NORMAL` (optimal: about 8 sec), `CALIBRATE_MODE::ACCURACY` (for a long time: about 38 sec). Without parameters autoCalibrate use parameter NORMAL.

> 

Reviewer: That’s why we should always start with documentation…

First, the terminology doesn’t play well: `setSensitivity` vs `autoCalibrate`. We calibrate sensitivity, thus a better name would be `calibrateSensitivity` or `optimizeSensitivity`. I find the latter slightly more clear.

Then, I see no reason to complicate things and introduce an extra level of indirection: the `enum CALIBRATE_MODE`. It means the single thing, the settling time of a single measurement. We could make it mean exactly this, don’t we? Useful presets could be defined as simple constants then:

```cpp
constexpr uint32_t OCTOLINER_ACCURACY_DRAFT = 120;
constexpr uint32_t OCTOLINER_ACCURACY_NORMAL = 480;
constexpr uint32_t OCTOLINER_ACCURACY_BEST = 2400;
```

Next, how the function behaves when it fails? For example, when the sensor is exposed to the void. This should be clearly defined. Now I see it will have “Integer divided by zero” error, not good. What should happen if the calibration fails after all? I think it’s up to the user to decide.

Putting all together, I suggest the following:

```cpp
### bool optimizeSensitivity();
### bool optimizeSensitivity(uint32_t measurementTimeMs);

Performs automatic sensitivity set up by making six subsequent measurements at different sensitivity levels.
The sensitivity level at which the contrast of the image ahead of Octoliner is the best wins. Make sure to put
the sensor over a contrast image like a line drawn on the surface before calling this method. For example, call
`optimizeSensitivity` in your `setup` and reset your robot to get the optimal sensitivity for the ride session.

Each measurement iteration has to settle down to make its result adequate. The time in milliseconds per
iteration is given through `measurementTimeMs` parameter. The following constants are defined as sane
choices:

- `OCTOLINER_ACCURACY_DRAFT` = 120 ms (2 sec total)
- `OCTOLINER_ACCURACY_NORMAL` = 480 ms (8 sec total, default parameter is omitted)
- `OCTOLINER_ACCURACY_BEST` = 2400 ms (38 sec total)

The method returns `true` if calibrated successfully. Use `getSensitivity` if you would like to know the value set.
If optimization fails, for example, when the sensor is exposed to void, the method returns `false`. In the case of failure, the currently set sensitivity level is left intact.

### uint8_t getSensitivity() const;

Returns the current sensitivity level previously set with `setSensitivity` or `optimizeSensitivity`.
```

Author: Yes, it is better then my version. I will rewrite code. But...where is 'Now I see it will have “Integer divided by zero” error'? Don`t look it.

Reviewer: Oops, missed the line:

```cpp
downCnt = (downCnt == 0) ? 1 : downCnt;
```

Indeed, it will never divide by 0;

Author: This code removed

If the current sensor reading does not allow understanding of the line position (some trash under the sensor or gone out of track), the previous successful estimation (which is cached) is returned. The starting value is `0.0`.

> ### `void optimizeSensitivityOnBlackField()`

Reviewer: ```suggestion
### `bool optimizeSensitivityOnBlack()`
```

It can fail. For example, when 8 sensors never settle at once. More examples of failure conditions are required.

Author: It can be only if octoliner is mechanicaly or electicaly broken. In other case all 8 will "1". On realy white (not black), it will be on _sensitivity = [120-150].

> Performs automatic sensitivity set up. Before optimization, place octoliner to track, all sensors other black line or field. Algorithm automatically adjust sensitivity for current track black level. Method set optimize value to class property `_sensitivity`.

Reviewer: ```suggestion
Performs automatic sensitivity set up. Before the optimization, place Octoliner over the track in a way where all its sensors point to the black line/field. The algorithm automatically adjusts the sensitivity for the current track black level.

The method returns `true` if succeed. If the calibration fails it returns `false`, for example, if la-la-la. In the case of failure the previous sensitivity level is left intact.

To get the sensitivity set, use `getSensitivity`.
```

> Measurement time is depends of black level. Typically value less then 1 sec, maximum is 3 sec.

Reviewer: ```suggestion
The measurement time depends on the black level. Typical value is less than 1 sec, the maximum is 3 sec.
```

### `uint8_t getSensitivity() const`

> Returns the current sensitivity level previously set with `setSensitivity` or `optimizeSensitivity`.

Reviewer: ```suggestion
Returns the current sensitivity level previously set with `setSensitivity` or `optimizeSensitivityOnBlack`.
```

If the current sensor reading does not allow understanding of the line position (some trash under the sensor or gone out of track), the previous successful estimation (which is cached) is returned. The starting value is `0.0`.

### `bool optimizeSensitivityOnBlack()`

> Performs automatic sensitivity set up. Before the optimization, place Octoliner over the track in a way where all its sensors point to the black line/field. The algorithm automatically adjusts the sensitivity for the current track black level.

Reviewer: ```suggestion
Performs automatic sensitivity set up. Before the optimization, place Octoliner over the track in a way where _all_ its sensors point to the black line/field. If it is physically impossible, point all the sensors to the void. The algorithm automatically adjusts the sensitivity for the current seen black level.
```

> The method returns `true` if succeed. If the calibration fails it returns `false`, for example, if one of sensors is mechanically or electrically broken. In the case of failure the previous sensitivity level is left intact.

Reviewer: ```suggestion
The method returns `true` if succeed. If the calibration fails it returns `false`, for example, if the sensors are placed over a contrast black/white or a completely white surface. In the case of failure, the previous sensitivity level is left intact.
```

The code should follow “the previous sensitivity level is left intact” behavior

To get the sensitivity set, use `getSensitivity`.

> The measurement time depends on the black level. Typical value is less than 1 sec, the maximum is 3 sec.

Reviewer: ```suggestion
The measurement time depends on the blackness level. The typical value is less than 1 sec, the maximum is 3 sec.
```
